apsuite/optics_analysis/coupling_correction.py

```diff
@@ -82,79 +85,78 @@ def get_ksl(self, model=None, skewidx=None):
 
     def coupling_corr_orbrespm_dispy(self,
                                      model,
-                                     matrix=None,
-                                     nsv=None, niter=10, tol=1e-6,
+                                     jacobian_matrix=None,
+                                     nsv=None, nr_max=10, tol=1e-6,
                                      res0=None):
         """Coupling correction with orbrespm.
 
         Calculates the pseudo-inverse of coupling correction matrix via SVD
         and minimizes the residue vector [Mxy, Myx, Etay].
         """
-        if matrix is None:
-            matrix = self.calc_coupling_matrix(model)
-        u, s, v = np.linalg.svd(matrix, full_matrices=False)
-        inv_s = 1/s
-        inv_s[np.isnan(inv_s)] = 0
-        inv_s[np.isinf(inv_s)] = 0
+        if jacobian_matrix is None:
+            jmat = self.calc_jacobian_matrix(model)
+        umat, smat, vmat = np.linalg.svd(jmat, full_matrices=False)
+        ismat = 1/smat
+        ismat[np.isnan(ismat)] = 0
+        ismat[np.isinf(ismat)] = 0
         if nsv is not None:
-            inv_s[nsv:] = 0
-        inv_s = np.diag(inv_s)
-        inv_matrix = -np.dot(np.dot(v.T, inv_s), u.T)
+            ismat[nsv:] = 0
+        ismat = np.diag(ismat)
+        ijmat = -np.dot(np.dot(vmat.T, ismat), umat.T)
         if res0 is None:
             res = self._get_coupling_residue(model)
         else:
             res = res0
-        bestfm = CouplingCorr.get_fm(res)
-        ksl0 = self.get_ksl(model)
-        ksl = ksl0
-
-        for i in range(niter):
-            dksl = np.dot(inv_matrix, res)
-            ksl += np.reshape(dksl, (-1, 1))
-            model = self.set_ksl(model=model, ksl=ksl)
+        bestfigm = CouplingCorr.get_figm(res)
+        if bestfigm < tol:
+            return CouplingCorr.CORR_STATUS.Sucess
+
+        for _ in range(nr_max):
+            dksl = np.dot(ijmat, res)
+            self._set_delta_ksl(model=model, delta_ksl=dksl)
             res = self._get_coupling_residue(model)
-            fm = CouplingCorr.get_fm(res)
-            diff_fm = np.abs(bestfm - fm)
-            print(i, bestfm)
-            if fm < bestfm:
-                bestfm = fm
-            if diff_fm < tol:
+            figm = CouplingCorr.get_figm(res)
+            diff_figm = np.abs(bestfigm - figm)
+            if figm < bestfigm:
+                bestfigm = figm
+            if diff_figm < tol:
                 break
-        print('done!')
-        return model
+        else:
+            return CouplingCorr.CORR_STATUS.Fail
+        return CouplingCorr.CORR_STATUS.Sucess
 
     def coupling_correction(self,
                             model,
-                            matrix=None, nsv=None, niter=10, tol=1e-6,
+                            jacobian_matrix=None,
+                            nsv=None, nr_max=10, tol=1e-6,
                             res0=None,
                             method='orbrespm'):
```

Comment:
I think `method` should be an object property, not an input argument.

Reply:
ok.

Reply:
see comment below